Sources/Purchasing/StoreKitAbstractions/OfferCodeRedemptionSheetPresenter.swift

```diff
@@ -0,0 +1,102 @@
+//
+//  Copyright RevenueCat Inc. All Rights Reserved.
+//
+//  Licensed under the MIT License (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      https://opensource.org/licenses/MIT
+//
+//  OfferCodeRedemptionSheetPresenter.swift
+//
+//  Created by Will Taylor on 10/17/24.
+
+import Foundation
+import StoreKit
+
+@objc protocol OfferCodeRedemptionSheetPresenterType: Sendable {
+
+    #if (os(iOS) || VISION_OS) && !targetEnvironment(macCatalyst)
+    @available(iOS 14.0, *)
+    @available(tvOS, unavailable)
+    @available(macOS, unavailable)
+    @available(macCatalyst, unavailable)
+    func presentCodeRedemptionSheet(windowScene: UIWindowScene) async throws
+    #endif
+}
+
+final internal class OfferCodeRedemptionSheetPresenter: OfferCodeRedemptionSheetPresenterType, Sendable {
+
+    private let paymentQueue: SKPaymentQueue
+    private let isiOSAppOnMac: Bool
+    private let osMajorVersion: Int
+
+    init(
+        paymentQueue: SKPaymentQueue = .default(),
+        isiOSAppOnMac: Bool = {
+            if #available(
+                iOS 14.0,
+                iOSApplicationExtension 14.0,
+                macOS 11.0,
+                watchOS 7.0,
+                tvOS 14.0,
+                *
+            ) {
+                return ProcessInfo().isiOSAppOnMac
+            } else {
+                return false
+            }
+        }(),
+        osMajorVersion: Int = {
+            return ProcessInfo().operatingSystemVersion.majorVersion
+        }()
+    ) {
+        self.paymentQueue = paymentQueue
+        self.isiOSAppOnMac = isiOSAppOnMac
+        self.osMajorVersion = osMajorVersion
+    }
+
+    #if (os(iOS) || VISION_OS) && !targetEnvironment(macCatalyst)
+    @available(iOS 14.0, *)
+    @available(tvOS, unavailable)
+    @available(macOS, unavailable)
+    @available(macCatalyst, unavailable)
+    func presentCodeRedemptionSheet(
+        windowScene: UIWindowScene
+    ) async throws {
+        if isiOSAppOnMac {
+            Logger.warn(Strings.storeKit.not_displaying_offer_code_redemption_sheet_because_ios_app_on_macos)
+            return
+        } else if osMajorVersion < 16 {
+            // .presentOfferCodeRedeemSheet(in: windowScene) isn't available in iOS <16, so fall back
+            // to the SK1 implementation
+            self.sk1PresentCodeRedemptionSheet()
+            return
+        }
+
+        if #available(iOS 16.0, iOSApplicationExtension 16.0, *) {
+            try await AppStore.presentOfferCodeRedeemSheet(in: windowScene)
+        } else {
+            // This case should be covered by the above OS check, but we'll include here
+            // since it's a possible code case
```

Comment:
Are we not taking into account SK1 vs SK2 preference here? 

I imagine it shouldn't make too much of a difference, but given that we know the systems are _eventually_ consistent and even API availability isn't the same, they might actually be showing two different sheets that look the same but go to SK1 / SK2 first. 

If that's the case, we should probably make sure that we're only using one SK version at a time

Reply:
Great point! I'll modify this to take the SK preference into account 😄